File: problem/boj/12865.js
const filePath =
  process.platform === "linux"
    ? "/dev/stdin"
    : "/Users/jeonsiwon/Desktop/coding-interview/code/boj/input.txt";
const input = require("fs")
  .readFileSync(filePath, "utf-8")
  .toString()
  .trim()
  .split("\n");
(function solution() {
  const [N, K] = input[0].split(" ").map(Number);
  const W = Array.from({ length: N + 1 }, () => 0);
  const V = Array.from({ length: N + 1 }, () => 0);
  const dp = Array.from({ length: 100001 }, () => 0);
  for (let i = 1; i <= N; i++) {
    const [w, v] = input[i].split(" ").map(Number);
    W[i] = w;
    V[i] = v;
  }
  for (let i = 1; i < W.length; i++) {
    for (let j = K; j >= W[i]; j--) {
      dp[j] = Math.max(dp[j - W[i]] + V[i], dp[j]);
    }
  }
  console.log(dp[K]);
})();

File: problem/boj/12865.cpp
#include <iostream>
using namespace std;

int N,K;
int V[101]={0,};
int W[101]={0,};
int dp[101][100001] ={0,};

void input(){
  cin >> N >> K;
  for(int i=1;i<=N;i++) cin >> W[i] >> V[i]; 
}

int main(void){
  ios::sync_with_stdio(false);
  cin.tie(0);
  cout.tie(0);
  freopen("input.txt","r",stdin);
  input();

  // operation
  for(int i=1;i<=N;i++){
    for(int j=1; j<=K;j++){
      if(j>=W[i])dp[i][j] = max(dp[i-1][j-W[i]]+V[i] , dp[i-1][j]);
      else dp[i][j] = dp[i-1][j];
    }
  } 
  
  cout << dp[N][K];
}